### ISSUE TITLE
[2.x] Unable to create distributed plugin (Class "filament" does not exist)

### ISSUE THREAD

**Issue by danilopolani** [laravel-filament/filament#805]
### Package

filament/filament

### Package Version

v2.0.0

### Laravel Version

v8.75.0

### PHP Version

8.0.3

### Bug description

Trying to register a Plugin extending the `PluginServiceProvider` class throws an error:

> Target class [filament] does not exist.

And

> [object] (ReflectionException(code: -1): Class \"filament\" does not exist

### Steps to reproduce

1. Create a distributed Plugin
2. Try to register it in a fresh Laravel application with Filament

Or, otherwise, you can use the package I'm trying to setup (still WIP 👀 ): https://github.com/danilopolani/filament-plausible-widget/tree/feat/v2

### Relevant log output

```shell
Target class [filament] does not exist. {"exception":"[object] (Illuminate\\Contracts\\Container\\BindingResolutionException(code: 0): Target class [filament] does not exist. at D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php:879)
[stacktrace]
#0 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php(758): Illuminate\\Container\\Container->build('filament')
#1 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Application.php(851): Illuminate\\Container\\Container->resolve('filament', Array, true)
#2 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php(694): Illuminate\\Foundation\\Application->resolve('filament', Array)
#3 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Application.php(836): Illuminate\\Container\\Container->make('filament', Array)
#4 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php(1423): Illuminate\\Foundation\\Application->make('filament')
#5 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(198): Illuminate\\Container\\Container->offsetGet('filament')
#6 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(166): Illuminate\\Support\\Facades\\Facade::resolveFacadeInstance('filament')
#7 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(255): Illuminate\\Support\\Facades\\Facade::getFacadeRoot()
#8 D:\\laravel\\vendor\\filament\\filament\\src\\PluginServiceProvider.php(33): Illuminate\\Support\\Facades\\Facade::__callStatic('registerPages', Array)
#9 D:\\laravel\\vendor\\spatie\\laravel-package-tools\\src\\PackageServiceProvider.php(36): Filament\\PluginServiceProvider->packageRegistered()
#10 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Application.php(682): Spatie\\LaravelPackageTools\\PackageServiceProvider->register()
#11 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\ProviderRepository.php(75): Illuminate\\Foundation\\Application->register(Object(DaniloPolani\\FilamentPlausibleWidget\\WidgetServiceProvider))
#12 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Application.php(659): Illuminate\\Foundation\\ProviderRepository->load(Array)
#13 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Bootstrap\\RegisterProviders.php(17): Illuminate\\Foundation\\Application->registerConfiguredProviders()
#14 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Application.php(237): Illuminate\\Foundation\\Bootstrap\\RegisterProviders->bootstrap(Object(Illuminate\\Foundation\\Application))
#15 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Console\\Kernel.php(310): Illuminate\\Foundation\\Application->bootstrapWith(Array)
#16 D:\\laravel\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Console\\Kernel.php(127): Illuminate\\Foundation\\Console\\Kernel->bootstrap()
#17 D:\\laravel\\artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#18 {main}
```
```

**Comment by danharrin** [comment]
Any clue why this isn't happening? Doesn't sound like the `FilamentManager` class is registered to the container, so is `FilamentServiceProvider` even being loaded?

**Comment by danilopolani** [comment]
> Any clue why this isn't happening? Doesn't sound like the `FilamentManager` class is registered to the container, so is `FilamentServiceProvider` even being loaded?

I was hoping you had the answer haha, I'm gonna try to debug it tomorrow. Are you able to reproduce this?

**Comment by danharrin** [comment]
> > Any clue why this isn't happening? Doesn't sound like the `FilamentManager` class is registered to the container, so is `FilamentServiceProvider` even being loaded?
> 
> I was hoping you had the answer haha, I'm gonna try to debug it tomorrow. Are you able to reproduce this?

Need to review all PRs first, then working through these issues. So you'll probably end up debugging before I do :)

**Comment by danharrin** [comment]
I haven't tried replicating yet, FYI.

**Comment by dllobell** [comment]
Hey!, I had the same issue today and found out that my PluginServiceProvider was registered before the main FilamentServiceProvider, and since the PluginServiceProvider `packageRegistered` hook uses the Filament facade to register pages, resources, etc., but the container binding does not exist yet then it throws the exception.

A workaround is to override the `packageRegistered` method and leave it empty and call the parent method from the `bootingPackage` or `packageBooted` hooks. Like this:

```php
    public function packageRegistered(): void
    {
    }

    public function packageBooted(): void
    {
        parent::packageRegistered();

        parent::packageBooted();
    }
```

Maybe we should move the code from the `packageRegistered` hook to another one from the PackageServiceProvider

**Comment by danilopolani** [comment]
> Hey!, I had the same issue today and found out that my PluginServiceProvider was registered before the main FilamentServiceProvider, and since the PluginServiceProvider `packageRegistered` hook uses the Filament facade to register pages, resources, etc., but the container binding does not exist yet then it throws the exception.
> 
> A workaround is to override the `packageRegistered` method and leave it empty and call the parent method from the `bootingPackage` or `packageBooted` hooks. Like this:
> 
> ```
>     public function packageRegistered(): void
>     {
>     }
> 
>     public function packageBooted(): void
>     {
>         parent::packageRegistered();
> 
>         parent::packageBooted();
>     }
> ```
> 
> Maybe we should move the code from the `packageRegistered` hook to another one from the PackageServiceProvider

I see, but `register` and `boot` are different methods and should not be mixed. Starting from your point a quick workaround is to register Filament from our plugins if it's not yet registered:

```php
public function packageRegistered(): void
{
    $this->app->singletonIf('filament', fn (): FilamentManager => new FilamentManager());

    parent::packageRegistered();
}
```

Maybe this could be implemented directly in the `PluginServiceProvider` before any other statement? cc @danharrin

**Comment by danharrin** [comment]
@danilopolani that looks like a good solution, what is it that dictates the order of service providers in Laravel? Is it alphabetical or something?

**Comment by danilopolani** [comment]
> @danilopolani that looks like a good solution, what is it that dictates the order of service providers in Laravel? Is it alphabetical or something?

Yea definitely, it seems like [Laravel is reading](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/PackageManifest.php#L122) from `composer/installed.json` that contains a list of alphabetically-ordered installed packages. I think we can't do much about it 🤷

**Comment by danharrin** [comment]
That's a shame. Do you want to open a PR to add this to core, or shall I?

**Comment by danilopolani** [comment]
> That's a shame. Do you want to open a PR to add this to core, or shall I?

I'm gonna take it, don't worry 😃

**Comment by danharrin** [comment]
Fixed by #918.
